keras_nlp/models/albert/albert_backbone.py

```diff
@@ -0,0 +1,264 @@
+            [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+        "padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+    }
+
+    # Randomly initialized ALBERT encoder
+    model = keras_nlp.models.AlbertBackbone(
+        vocabulary_size=30000,
+        num_layers=12,
+        num_heads=12,
+        num_groups=1,
+        num_inner_repetitions=1,
+        embedding_dim=128,
+        hidden_dim=768,
+        intermediate_dim=3072,
+        max_sequence_length=12,
+    )
+    output = model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        num_groups,
+        num_inner_repetitions,
+        embedding_dim,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.0,
+        max_sequence_length=512,
+        num_segments=2,
+        **kwargs,
+    ):
+
+        # Index of classification token in the vocabulary
+        cls_token_index = 0
+        # Inputs
+        token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="token_ids"
+        )
+        segment_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="segment_ids"
+        )
+        padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+
+        # Embed tokens, positions, and segment ids.
+        token_embedding_layer = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=embedding_dim,
+            embeddings_initializer=albert_kernel_initializer(),
+            name="token_embedding",
+        )
+        token_embedding = token_embedding_layer(token_id_input)
+        position_embedding = PositionEmbedding(
+            initializer=albert_kernel_initializer(),
+            sequence_length=max_sequence_length,
+            name="position_embedding",
+        )(token_embedding)
+        segment_embedding = keras.layers.Embedding(
+            input_dim=num_segments,
+            output_dim=embedding_dim,
+            embeddings_initializer=albert_kernel_initializer(),
+            name="segment_embedding",
+        )(segment_id_input)
+
+        # Sum, normalize and apply dropout to embeddings.
+        x = keras.layers.Add()(
+            (token_embedding, position_embedding, segment_embedding)
+        )
+        x = keras.layers.LayerNormalization(
+            name="embeddings_layer_norm",
+            axis=-1,
+            epsilon=1e-12,
+            dtype=tf.float32,
+        )(x)
+        x = keras.layers.Dropout(
+            dropout,
+            name="embeddings_dropout",
+        )(x)
+
+        # Project the embedding to `hidden_dim`.
+        x = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=albert_kernel_initializer(),
+            name="embedding_projection",
+        )(x)
+
+        def get_group_layer(group_idx):
```

Comment:
Ah, this may be my bad, I recommended this (it's a style used in other parts of Keras somewhat frequently). Ideally, we could just use a `keras.Sequential` here, which would be super readable. But there is no way to pass a padding mask to a sequential.

Reply:
@jbischof, yeah, that's what I'd initially done in the first few commits. But we want to avoid making extra layers whenever possible, discussed this with @mattdangerw. And secondly, we will have to pass a bunch of args to the layer in the for loop, which is something we can avoid.

This snippet, which is succinct and small:
```
        for group_idx in range(num_groups):
            # Define the group. A group in ALBERT terminology is any number of
            # repeated attention and FFN blocks.
            group_layer = get_group_layer(group_idx)

            # Assume num_layers = 8, num_groups = 5. Then, the order of group
            # calls will be 0, 0, 1, 1, 2, 3, 3, 4.
            while int(layer_idx / num_calls_per_group) == group_idx:
                x = group_layer(x, padding_mask=padding_mask)
                layer_idx += 1
```

, will become this:
```
        for group_idx in range(num_groups):
            # Define the group. A group in ALBERT terminology is any number of
            # repeated attention and FFN blocks.
            group_layer = GroupLayer(
                group_idx, 
                num_heads=num_heads,
                intermediate_dim=intermediate_dim,
                dropout=dropout,
                ...
            )

            # Assume num_layers = 8, num_groups = 5. Then, the order of group
            # calls will be 0, 0, 1, 1, 2, 3, 3, 4.
            while int(layer_idx / num_calls_per_group) == group_idx:
                x = group_layer(x, padding_mask=padding_mask)
                layer_idx += 1
```
, or something similar, anyway. 

On second thoughts, doesn't look too bad, but yeah, the first point still stands.

Reply:
OK, so be it

Reply:
There is this, which has the advantage of showing up in a summary/plot_model call as a single entity.

```python
def get_group_layer(group_idx):
    """Stack `num_inner_repetitions` encoder block as a single layer."""
    outputs = inputs = keras.Input(shape=(None, hidden_dim))
    for _ in range(num_inner_repetitions):
        outputs = TransformerEncoder(
            num_heads=num_heads,
            intermediate_dim=intermediate_dim,
            activation=lambda x: keras.activations.gelu(
                x, approximate=True
            ),
            dropout=dropout,
            kernel_initializer=albert_kernel_initializer(),
        )(outputs, padding_mask=padding_mask)
    return keras.Model(
        (inputs, padding_mask), (outputs), name=f"group_{group_idx}",
    )
```

It needs to be called as a tuple `group_layer((x, padding_mask))`.

I'm also totally happy to just quit and remove `num_groups` and `num_inner_repetitions` entirely if we want to :imp:

Reply:
> I'm also totally happy to just quit and remove `num_groups` and `num_inner_repetitions` entirely if we want to 😈

![image](https://user-images.githubusercontent.com/31360468/211127836-996083c4-6d0a-46b3-bc60-01de8afa6caf.png)

Reply:
Can do the functional approach! Currently, this is how `model.summary()` looks. I feel this isn't too bad, especially because the arguments are sort of confusing. This helps clarify which arg does what, especially `num_groups` and `num_inner_repetitions`.

```
Model: "albert_backbone"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 token_ids (InputLayer)         [(None, None)]       0           []                               
                                                                                                  
 token_embedding (Embedding)    (None, None, 128)    3840000     ['token_ids[0][0]']              
                                                                                                  
 segment_ids (InputLayer)       [(None, None)]       0           []                               
                                                                                                  
 position_embedding (PositionEm  (None, None, 128)   65536       ['token_embedding[0][0]']        
 bedding)                                                                                         
                                                                                                  
 segment_embedding (Embedding)  (None, None, 128)    256         ['segment_ids[0][0]']            
                                                                                                  
 add (Add)                      (None, None, 128)    0           ['token_embedding[0][0]',        
                                                                  'position_embedding[0][0]',     
                                                                  'segment_embedding[0][0]']      
                                                                                                  
 embeddings_layer_norm (LayerNo  (None, None, 128)   256         ['add[0][0]']                    
 rmalization)                                                                                     
                                                                                                  
 embeddings_dropout (Dropout)   (None, None, 128)    0           ['embeddings_layer_norm[0][0]']  
                                                                                                  
 embedding_projection (Dense)   (None, None, 768)    99072       ['embeddings_dropout[0][0]']     
                                                                                                  
 padding_mask (InputLayer)      [(None, None)]       0           []                               
                                                                                                  
 group_0_transformer_layer_0 (T  (None, None, 768)   7087872     ['embedding_projection[0][0]',   
 ransformerEncoder)                                               'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[0][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[1][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[2][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[3][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[4][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[5][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[6][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[7][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[8][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[9][
                                                                 0]',                             
                                                                  'padding_mask[0][0]',           
                                                                  'group_0_transformer_layer_0[10]
                                                                 [0]',                            
                                                                  'padding_mask[0][0]']           
                                                                                                  
 tf.__operators__.getitem (Slic  (None, 768)         0           ['group_0_transformer_layer_0[11]
 ingOpLambda)                                                    [0]']                            
                                                                                                  
 pooled_dense (Dense)           (None, 768)          590592      ['tf.__operators__.getitem[0][0]'
                                                                 ]                                
                                                                                                  
==================================================================================================
Total params: 11,683,584
Trainable params: 11,683,584
Non-trainable params: 0
__________________________________________________________________________________________________
```